stan/math/rev/functor/cvodes_integrator.hpp

```diff
@@ -75,116 +183,174 @@ class cvodes_integrator {
    * @throw <code>std::invalid_argument</code> if arguments are the wrong
    *   size or tolerances or max_num_steps are out of range.
    */
-  template <typename F, typename T_initial, typename T_param, typename T_t0,
-            typename T_ts>
-  std::vector<std::vector<return_type_t<T_initial, T_param, T_t0, T_ts>>>
-  integrate(const char* function_name, const F& f,
-            const std::vector<T_initial>& y0, const T_t0& t0,
-            const std::vector<T_ts>& ts, const std::vector<T_param>& theta,
-            const std::vector<double>& x, const std::vector<int>& x_int,
-            std::ostream* msgs, double relative_tolerance,
-            double absolute_tolerance,
-            long int max_num_steps) {  // NOLINT(runtime/int)
-    using initial_var = stan::is_var<T_initial>;
-    using param_var = stan::is_var<T_param>;
-
-    const double t0_dbl = value_of(t0);
-    const std::vector<double> ts_dbl = value_of(ts);
-
-    check_finite(function_name, "initial state", y0);
-    check_finite(function_name, "initial time", t0_dbl);
-    check_finite(function_name, "times", ts_dbl);
-    check_finite(function_name, "parameter vector", theta);
-    check_finite(function_name, "continuous data", x);
-    check_nonzero_size(function_name, "times", ts);
-    check_nonzero_size(function_name, "initial state", y0);
-    check_ordered(function_name, "times", ts_dbl);
-    check_less(function_name, "initial time", t0_dbl, ts_dbl[0]);
-    if (relative_tolerance <= 0) {
-      invalid_argument(function_name, "relative_tolerance,", relative_tolerance,
-                       "", ", must be greater than 0");
-    }
-    if (absolute_tolerance <= 0) {
-      invalid_argument(function_name, "absolute_tolerance,", absolute_tolerance,
-                       "", ", must be greater than 0");
+  cvodes_integrator(const F& f,
+                    const Eigen::Matrix<T_y0, Eigen::Dynamic, 1>& y0,
+                    const T_t0& t0, const std::vector<T_ts>& ts,
+                    double relative_tolerance, double absolute_tolerance,
+                    bool include_sensitivities_in_errors,
+                    long int max_num_steps,  // NOLINT(runtime/int)
+                    std::ostream* msgs, const T_Args&... args)
+      : f_(f),
+        y0_(y0.unaryExpr([](const T_y0& val) { return T_y0_t0(val); })),
+        t0_(t0),
+        ts_(ts),
+        args_tuple_(args...),
+        value_of_args_tuple_(value_of(args)...),
+        N_(y0.size()),
+        msgs_(msgs),
+        relative_tolerance_(relative_tolerance),
+        absolute_tolerance_(absolute_tolerance),
+        include_sensitivities_in_errors_(include_sensitivities_in_errors),
+        max_num_steps_(max_num_steps),
+        y0_vars_(count_vars(y0_)),
+        args_vars_(count_vars(args...)),
+        coupled_ode_(f, y0_, msgs, args...),
+        coupled_state_(coupled_ode_.initial_state()) {
+    const char* fun = "cvodes_integrator::integrate";
```

Comment:
hmm? This is the constructor, right? I am not sure if cvodes_integrator::integrate is then the right thing to call this `fun`.